File: experiment/backends/android/dazzle/DazzlePipelineContextManager.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package dev.dazzle.experiment

import dev.dazzle.sdk.RespParser
import dev.dazzle.sdk.RespValue
import dev.dazzle.sdk.StreamKey
import dev.dazzle.sdk.Dazzle
import dev.dazzle.sdk.DazzleServer

/**
 * Optimized Valkey [StorageBackend] that uses directPipeline to batch
 * all writes AND reads into minimal pipe round-trips.
 *
 * The naive ValkeyContextManager issues 6-8 separate directCommand calls
 * per buildContextBlock, each crossing the event-loop pipe independently.
 * This version batches:
 *   - Writes: all ingest ops in 1 pipeline (1 round-trip instead of 6-8)
 *   - Reads: all context queries in 1 pipeline (1 round-trip instead of 6-8)
 *
 * This demonstrates that Valkey's directPipeline capability directly
 * reduces retrieval latency by 3-5× with zero application logic change.
 */
class DazzlePipelineContextManager : StorageBackend {

    override val backendName: String = "Dazzle-Pipeline"

    private val dazzle: Dazzle = DazzleServer.client()
    private val readings  = dazzle.stream("sensor:readings")
    private val stats     = dazzle.hash("sensor:stats")
    private val anomalies = dazzle.sortedSet("sensor:anomalies")
    private val decisions = dazzle.list("agent:decisions")

    override fun flush() {
        readings.deleteKey()
        stats.delete()
        anomalies.deleteKey()
        decisions.deleteKey()
        for (i in 0..9) dazzle.hash("agent:checkpoint:$i").delete()
    }

    // ── Ingest (batched pipeline) ─────────────────────────────────────────

    override fun ingest(reading: SensorReading) {
        val commands = mutableListOf<List<String>>()

        commands.add(listOf(
            "XADD", "sensor:readings", "MAXLEN", "~", "200", "*",
            "temp", reading.tempC.toString(),
            "humidity", reading.humidity.toString(),
            "minute", reading.minute.toString(),
            "anomalous", if (reading.anomalous) "1" else "0",
        ))
        commands.add(listOf("HINCRBYFLOAT", "sensor:stats", "temp_sum", reading.tempC.toString()))
        commands.add(listOf("HINCRBY", "sensor:stats", "count", "1"))
        commands.add(listOf("HSET", "sensor:stats", "latest_temp", reading.tempC.toString()))
        commands.add(listOf("HSET", "sensor:stats", "latest_minute", reading.minute.toString()))

        if (reading.anomalous) {
            commands.add(listOf("ZADD", "sensor:anomalies", reading.minute.toString(), reading.minute.toString()))
            commands.add(listOf("HINCRBY", "sensor:stats", "anomaly_count", "1"))
        }

        DazzleServer.directPipeline(commands)

        // Min/max needs conditional (can't fully pipeline)
        val curMin = stats.get("min_temp")?.toDoubleOrNull()
        if (curMin == null || reading.tempC < curMin) {
            stats.set("min_temp", reading.tempC.toString())
        }
        val curMax = stats.get("max_temp")?.toDoubleOrNull()
        if (curMax == null || reading.tempC > curMax) {
            stats.set("max_temp", reading.tempC.toString())
        }
    }

    // ── Context block (batched read pipeline) ─────────────────────────────

    override fun buildContextBlock(currentMinute: Int, windowMinutes: Int): String {
        val windowStart = maxOf(0, currentMinute - windowMinutes)

        // Batch ALL reads in one pipeline (1 pipe round-trip instead of 6-8)
        val replies = DazzleServer.directPipeline(listOf(
            listOf("XREVRANGE", "sensor:readings", "+", "-", "COUNT", "10"),
            listOf("HMGET", "sensor:stats", "count", "temp_sum", "min_temp", "max_temp", "anomaly_count"),
            listOf("ZRANGEBYSCORE", "sensor:anomalies", windowStart.toString(), currentMinute.toString()),
        ))

        val lines = mutableListOf<String>()

        // Parse recent readings from XREVRANGE reply
        val readingsReply = replies.getOrNull(0)
        if (readingsReply != null) {
            try {
                val parsed = RespParser.parse(readingsReply)
                val entries = parsed.asArray()
                val recentTemps = mutableListOf<Double>()
                for (item in entries.reversed()) {
                    val pair = item.asArray()
                    if (pair.size >= 2) {
                        val fields = pair[1].asArray()
                        var i = 0
                        while (i < fields.size - 1) {
                            if (fields[i].asBulkOrNull() == "temp") {
                                fields[i + 1].asBulkOrNull()?.toDoubleOrNull()?.let { recentTemps.add(it) }
                            }
                            i += 2
                        }
                    }
                }
                if (recentTemps.isNotEmpty()) {
                    val formatted = recentTemps.joinToString(", ") { String.format("%.1f", it) }
                    lines += "Last ${recentTemps.size} temperatures (oldest→newest, °C): $formatted"
                    lines += "Recent trend: ${computeTrend(recentTemps)}"
                }
            } catch (_: Exception) { /* skip if parse fails */ }
        }

        // Parse stats from HMGET reply
        val statsReply = replies.getOrNull(1)
        if (statsReply != null) {
            try {
                val parsed = RespParser.parse(statsReply).asArray()
                val count = parsed.getOrNull(0)?.asBulkOrNull()?.toIntOrNull()
                if (count != null && count > 0) {
                    val sum = parsed.getOrNull(1)?.asBulkOrNull()?.toDoubleOrNull() ?: 0.0
                    val minT = parsed.getOrNull(2)?.asBulkOrNull()?.toDoubleOrNull() ?: 0.0
                    val maxT = parsed.getOrNull(3)?.asBulkOrNull()?.toDoubleOrNull() ?: 0.0
                    val anomCnt = parsed.getOrNull(4)?.asBulkOrNull()?.toIntOrNull() ?: 0
                    lines += "Aggregate over $count readings: " +
                        "avg=${String.format("%.1f", sum / count)}°C, " +
                        "min=${String.format("%.1f", minT)}°C, " +
                        "max=${String.format("%.1f", maxT)}°C"
                    lines += "Total anomalies detected so far: $anomCnt"
                }
            } catch (_: Exception) { /* skip */ }
        }

        // Parse anomalies from ZRANGEBYSCORE reply
        val anomReply = replies.getOrNull(2)
        if (anomReply != null) {
            try {
                val parsed = RespParser.parse(anomReply).asArray()
                val windowAnomalies = parsed.mapNotNull { it.asBulkOrNull() }
                if (windowAnomalies.isEmpty()) {
                    lines += "No anomalies in the last $windowMinutes minutes."
                } else {
                    lines += "Anomalous time indices in the last $windowMinutes minutes " +
                        "(minute numbers, not temperatures): [${windowAnomalies.joinToString(", ")}]"
                }
            } catch (_: Exception) {
                lines += "No anomalies in the last $windowMinutes minutes."
            }
        }

        return lines.joinToString("\n")
    }

    override fun buildSynthesisContext(): String {
        val replies = DazzleServer.directPipeline(listOf(
            listOf("HMGET", "sensor:stats", "count", "temp_sum", "min_temp", "max_temp", "anomaly_count"),
            listOf("ZRANGEBYSCORE", "sensor:anomalies", "0", "99999"),
            listOf("LRANGE", "agent:decisions", "0", "-1"),
        ))

        val lines = mutableListOf<String>()

        // Stats
        val statsReply = replies.getOrNull(0)
        if (statsReply != null) {
            try {
                val parsed = RespParser.parse(statsReply).asArray()
                val count = parsed.getOrNull(0)?.asBulkOrNull()?.toIntOrNull()
                if (count != null && count > 0) {
                    val sum = parsed.getOrNull(1)?.asBulkOrNull()?.toDoubleOrNull() ?: 0.0
                    val minT = parsed.getOrNull(2)?.asBulkOrNull()?.toDoubleOrNull() ?: 0.0
                    val maxT = parsed.getOrNull(3)?.asBulkOrNull()?.toDoubleOrNull() ?: 0.0
                    val anomCnt = parsed.getOrNull(4)?.asBulkOrNull()?.toIntOrNull() ?: 0
                    lines += "=== Full Session Stats ==="
                    lines += "Total readings: $count"
                    lines += "Temperature range: ${String.format("%.1f", minT)}°C to " +
                        "${String.format("%.1f", maxT)}°C " +
                        "(avg ${String.format("%.1f", sum / count)}°C)"
                    lines += "Total anomalies detected: $anomCnt"
                }
            } catch (_: Exception) { /* skip */ }
        }

        // All anomaly minutes
        val anomReply = replies.getOrNull(1)
        if (anomReply != null) {
            try {
                val anomalies = RespParser.parse(anomReply).asArray()
                    .mapNotNull { it.asBulkOrNull() }
                if (anomalies.isNotEmpty()) {
                    lines += "Anomalous time indices (minute numbers, not temperatures): " +
                        "[${anomalies.joinToString(", ")}]"
                }
            } catch (_: Exception) { /* skip */ }
        }

        // Decisions
        val decReply = replies.getOrNull(2)
        if (decReply != null) {
            try {
                val decs = RespParser.parse(decReply).asArray()
                    .mapNotNull { it.asBulkOrNull() }
                if (decs.isNotEmpty()) {
                    lines += "=== Monitoring Agent Decisions ==="
                    for ((idx, decision) in decs.withIndex()) {
                        lines += "  Checkpoint ${idx + 1}: $decision"
                    }
                }
            } catch (_: Exception) { /* skip */ }
        }

        return lines.joinToString("\n")
    }

    override fun storeCheckpointDecision(
        index: Int,
        minute: Int,
        anomalyDetected: Boolean,
        severity: String,
        trend: String,
    ) {
        val decision = "anomaly=${if (anomalyDetected) "yes" else "no"} " +
            "severity=$severity trend=$trend"

        DazzleServer.directPipeline(listOf(
            listOf("HSET", "agent:checkpoint:$index",
                "minute", minute.toString(),
                "anomaly", if (anomalyDetected) "1" else "0",
                "severity", severity,
                "trend", trend),
            listOf("RPUSH", "agent:decisions", decision),
        ))
    }

    private fun computeTrend(temps: List<Double>): String {
        if (temps.size < 2) return "stable"
        val n = temps.size
        val meanX = (n - 1) / 2.0
        val meanY = temps.average()
        val num = (0 until n).sumOf { (it - meanX) * (temps[it] - meanY) }
        val den = (0 until n).sumOf { (it - meanX) * (it - meanX) }
        val slope = if (den != 0.0) num / den else 0.0
        return when {
            slope >  0.15 -> "increasing"
            slope < -0.15 -> "decreasing"
            else          -> "stable"
        }
    }

    // ── Footprint accounting ──────────────────────────────────────────────

    override val backendSizeMethod: String = "valkey:used_memory_dataset"

    override fun backendSizeBytes(): Long {
        val info = dazzle.server().info("memory")
        return info.field("Memory", "used_memory_dataset")?.toLongOrNull()
            ?: info.usedMemoryBytes
            ?: -1L
    }

    override fun backendSizeBreakdown(): Map<String, Long> =
        valkeyMemoryBreakdown(dazzle.server().info("memory"))
}

File: experiment/backends/ios/dazzle/DazzlePipelineContextManager.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

/// Optimized Valkey StorageBackend that uses directPipeline to batch
/// all writes AND reads into minimal pipe round-trips.
///
/// The naive DazzleContextManager issues 6-8 separate directCommand calls
/// per buildContextBlock. This version batches them via directPipelineArgs
/// (in-process, pre-split argv — no space-splitting issues).
final class DazzlePipelineContextManager: StorageBackend {

    let backendName: String = "Dazzle-Pipeline"

    private let dazzle: Dazzle
    private let readings:  StreamKey
    private let stats:     HashKey
    private let anomalies: SortedSetKey
    private let decisions: ListKey

    init() {
        let client = DazzleServer.shared.client()
        self.dazzle    = client
        self.readings  = client.stream("sensor:readings")
        self.stats     = client.hash("sensor:stats")
        self.anomalies = client.sortedSet("sensor:anomalies")
        self.decisions = client.list("agent:decisions")
    }

    func flush() {
        _ = try? readings.deleteKey()
        _ = try? stats.deleteKey()
        _ = try? anomalies.deleteKey()
        _ = try? decisions.deleteKey()
        for i in 0...9 { _ = try? dazzle.hash("agent:checkpoint:\(i)").deleteKey() }
    }

    // MARK: - Ingest (batched pipeline)

    func ingest(_ reading: SensorReading) {
        var commands: [[String]] = []

        commands.append(["XADD", "sensor:readings", "MAXLEN", "~", "200", "*",
            "temp", String(reading.tempC),
            "humidity", String(reading.humidity),
            "minute", String(reading.minute),
            "anomalous", reading.anomalous ? "1" : "0"])
        commands.append(["HINCRBYFLOAT", "sensor:stats", "temp_sum", String(reading.tempC)])
        commands.append(["HINCRBY", "sensor:stats", "count", "1"])
        commands.append(["HSET", "sensor:stats", "latest_temp", String(reading.tempC)])
        commands.append(["HSET", "sensor:stats", "latest_minute", String(reading.minute)])

        if reading.anomalous {
            commands.append(["ZADD", "sensor:anomalies", String(reading.minute), String(reading.minute)])
            commands.append(["HINCRBY", "sensor:stats", "anomaly_count", "1"])
        }

        _ = DazzleServer.shared.directPipelineArgs(commands)

        // Min/max needs conditional (can't fully pipeline)
        let curMin = (try? stats.get("min_temp")).flatMap { $0 }.flatMap { Double($0) }
        if curMin == nil || reading.tempC < curMin! {
            _ = try? stats.set("min_temp", String(reading.tempC))
        }
        let curMax = (try? stats.get("max_temp")).flatMap { $0 }.flatMap { Double($0) }
        if curMax == nil || reading.tempC > curMax! {
            _ = try? stats.set("max_temp", String(reading.tempC))
        }
    }

    // MARK: - Context block (batched read pipeline)

    func buildContextBlock(currentMinute: Int, windowMinutes: Int = 20) -> String {
        let windowStart = max(0, currentMinute - windowMinutes)

        // Batch ALL reads in one pipeline
        let replies = DazzleServer.shared.directPipelineArgs([
            ["XREVRANGE", "sensor:readings", "+", "-", "COUNT", "10"],
            ["HMGET", "sensor:stats", "count", "temp_sum", "min_temp", "max_temp", "anomaly_count"],
            ["ZRANGEBYSCORE", "sensor:anomalies", String(windowStart), String(currentMinute)],
        ])

        var lines: [String] = []

        // Parse recent readings from XREVRANGE reply
        if let readingsReply = replies.first, !readingsReply.isEmpty {
            if let parsed = try? RespParser.parse(readingsReply) {
                let entries = parsed.asArray
                var recentTemps: [Double] = []
                for item in entries.reversed() {
                    let pair = item.asArray
                    if pair.count >= 2 {
                        let fields = pair[1].asArray
                        var i = 0
                        while i < fields.count - 1 {
                            if fields[i].asBulkOrNil == "temp" {
                                if let t = fields[i + 1].asBulkOrNil.flatMap({ Double($0) }) {
                                    recentTemps.append(t)
                                }
                            }
                            i += 2
                        }
                    }
                }
                if !recentTemps.isEmpty {
                    let formatted = recentTemps.map { String(format: "%.1f", $0) }.joined(separator: ", ")
                    lines.append("Last \(recentTemps.count) temperatures (oldest\u{2192}newest, \u{00B0}C): \(formatted)")
                    lines.append("Recent trend: \(computeTrend(recentTemps))")
                }
            }
        }

        // Parse stats from HMGET reply
        if replies.count > 1 {
            let statsReply = replies[1]
            if !statsReply.isEmpty, let parsed = try? RespParser.parse(statsReply) {
                let arr = parsed.asArray
                if let count = arr.first?.asBulkOrNil.flatMap({ Int($0) }), count > 0 {
                    let sum  = (arr.count > 1 ? arr[1].asBulkOrNil : nil).flatMap { Double($0) } ?? 0
                    let minT = (arr.count > 2 ? arr[2].asBulkOrNil : nil).flatMap { Double($0) } ?? 0
                    let maxT = (arr.count > 3 ? arr[3].asBulkOrNil : nil).flatMap { Double($0) } ?? 0
                    let anomCnt = (arr.count > 4 ? arr[4].asBulkOrNil : nil).flatMap { Int($0) } ?? 0
                    lines.append("Aggregate over \(count) readings: " +
                        "avg=\(String(format: "%.1f", sum / Double(count)))\u{00B0}C, " +
                        "min=\(String(format: "%.1f", minT))\u{00B0}C, " +
                        "max=\(String(format: "%.1f", maxT))\u{00B0}C")
                    lines.append("Total anomalies detected so far: \(anomCnt)")
                }
            }
        }

        // Parse anomalies from ZRANGEBYSCORE reply
        if replies.count > 2 {
            let anomReply = replies[2]
            if !anomReply.isEmpty, let parsed = try? RespParser.parse(anomReply) {
                let windowAnomalies = parsed.asArray.compactMap { $0.asBulkOrNil }
                if windowAnomalies.isEmpty {
                    lines.append("No anomalies in the last \(windowMinutes) minutes.")
                } else {
                    lines.append("Anomalous time indices in the last \(windowMinutes) minutes " +
                        "(minute numbers, not temperatures): [\(windowAnomalies.joined(separator: ", "))]")
                }
            } else {
                lines.append("No anomalies in the last \(windowMinutes) minutes.")
            }
        } else {
            lines.append("No anomalies in the last \(windowMinutes) minutes.")
        }

        return lines.joined(separator: "\n")
    }

    func buildSynthesisContext() -> String {
        let replies = DazzleServer.shared.directPipelineArgs([
            ["HMGET", "sensor:stats", "count", "temp_sum", "min_temp", "max_temp", "anomaly_count"],
            ["ZRANGEBYSCORE", "sensor:anomalies", "0", "99999"],
            ["LRANGE", "agent:decisions", "0", "-1"],
        ])

        var lines: [String] = []

        if let statsReply = replies.first, !statsReply.isEmpty,
           let parsed = try? RespParser.parse(statsReply) {
            let arr = parsed.asArray
            if let count = arr.first?.asBulkOrNil.flatMap({ Int($0) }), count > 0 {
                let sum  = (arr.count > 1 ? arr[1].asBulkOrNil : nil).flatMap { Double($0) } ?? 0
                let minT = (arr.count > 2 ? arr[2].asBulkOrNil : nil).flatMap { Double($0) } ?? 0
                let maxT = (arr.count > 3 ? arr[3].asBulkOrNil : nil).flatMap { Double($0) } ?? 0
                let anomCnt = (arr.count > 4 ? arr[4].asBulkOrNil : nil).flatMap { Int($0) } ?? 0
                lines.append("=== Full Session Stats ===")
                lines.append("Total readings: \(count)")
                lines.append("Temperature range: \(String(format: "%.1f", minT))\u{00B0}C to " +
                    "\(String(format: "%.1f", maxT))\u{00B0}C " +
                    "(avg \(String(format: "%.1f", sum / Double(count)))\u{00B0}C)")
                lines.append("Total anomalies detected: \(anomCnt)")
            }
        }

        if replies.count > 1 {
            let anomReply = replies[1]
            if !anomReply.isEmpty, let parsed = try? RespParser.parse(anomReply) {
                let anoms = parsed.asArray.compactMap { $0.asBulkOrNil }
                if !anoms.isEmpty {
                    lines.append("Anomalous time indices (minute numbers, not temperatures): " +
                        "[\(anoms.joined(separator: ", "))]")
                }
            }
        }

        if replies.count > 2 {
            let decReply = replies[2]
            if !decReply.isEmpty, let parsed = try? RespParser.parse(decReply) {
                let decs = parsed.asArray.compactMap { $0.asBulkOrNil }
                if !decs.isEmpty {
                    lines.append("=== Monitoring Agent Decisions ===")
                    for (idx, decision) in decs.enumerated() {
                        lines.append("  Checkpoint \(idx + 1): \(decision)")
                    }
                }
            }
        }

        return lines.joined(separator: "\n")
    }

    func storeCheckpointDecision(index: Int, minute: Int, anomalyDetected: Bool, severity: String, trend: String) {
        let decision = "anomaly=\(anomalyDetected ? "yes" : "no") severity=\(severity) trend=\(trend)"

        _ = DazzleServer.shared.directPipelineArgs([
            ["HSET", "agent:checkpoint:\(index)",
                "minute", String(minute),
                "anomaly", anomalyDetected ? "1" : "0",
                "severity", severity,
                "trend", trend],
            ["RPUSH", "agent:decisions", decision],
        ])
    }

    private func computeTrend(_ temps: [Double]) -> String {
        guard temps.count >= 2 else { return "stable" }
        let n = temps.count
        let meanX = Double(n - 1) / 2.0
        let meanY = temps.reduce(0, +) / Double(n)
        let num = (0..<n).reduce(0.0) { $0 + (Double($1) - meanX) * (temps[$1] - meanY) }
        let den = (0..<n).reduce(0.0) { $0 + pow(Double($1) - meanX, 2) }
        guard den != 0 else { return "stable" }
        let slope = num / den
        if slope > 0.15 { return "increasing" }
        if slope < -0.15 { return "decreasing" }
        return "stable"
    }

    // MARK: - Footprint accounting

    var backendSizeMethod: String { "valkey:used_memory_dataset" }

    func backendSizeBytes() -> Int64 {
        guard let info = DazzleServer.shared.directCommand("INFO memory") else { return -1 }
        return parseValkeyUsedMemoryDataset(info)
    }

    func backendSizeBreakdown() -> [String: Int64]? {
        guard let info = DazzleServer.shared.directCommand("INFO memory") else { return nil }
        return parseValkeyMemoryStats(info)
    }
}
